### PR TITLE
Fix handling of disconnected lingering nodes

### PR DESCRIPTION
**Pull request by codablock**
See individual commits.

### CODE REVIEW

**Comment by codablock** [comment]
Test failures were probably related due to timing changes in DisconnectNodes. Latest commit should fix it.

**Comment by codablock** [comment]
Turned out that the lingering that was recently added actually broke the fixes (76d7b17dcdb7fbdd66decd5250a3e26ef600efe5 and 2f1b3a34ab9694c40c184cb7302181eca59ba9cd) of `getpeerinfo` and `getconnectioncount` returning peers which are being disconnected. I changed this PR to revert these two commits and add a new fixes which should work better and independent from lingering.

The reason this was not spotted before is that the CPU burning which was initially fixed by this PR made `DisconnectNodes` force disconnect the peers on the first call already, while after fixing it would return and not disconnect yet.

File: src/net.cpp [hunk 1]
```diff
@@ -323,44 +323,44 @@ bool IsReachable(const CNetAddr& addr)
 }
 
 
-CNode* CConnman::FindNode(const CNetAddr& ip)
+CNode* CConnman::FindNode(const CNetAddr& ip, bool fExcludeDisconnecting)
 {
     LOCK(cs_vNodes);
     for (CNode* pnode : vNodes) {
-        if ((CNetAddr)pnode->addr == ip) {
+        if ((CNetAddr)pnode->addr == ip && (!fExcludeDisconnecting || !pnode->fDisconnect)) {
```
[the remark this answers is not in the document]
Comment (codablock) [reply src/net.cpp]: agree, but fixed readability with a slightly different fix

**Comment by codablock** [comment]
@UdjinM6 do you have logs for the failure you tried to fix? The link you posted has all tests passed. I'm wondering why this fix was needed, because the FindNode+GetNodeCount+GetNodeStats fixes should have fixed exactly what you're trying to fix there.

**Comment by UdjinM6** [comment]
Tests are failing right here, in this PR, see gitlab. The link I posted is the "proof" that tests pass when the patch is applied.

**Comment by codablock** [comment]
@UdjinM6 ok I think I understand why you had to do the fix. The problem was that node3 initiated the disconnect to node0, then after lingering closed the socket. At this time node3 has removed node0 from his node list, so all is good on that side. But node0 now just started the lingering, so it won't close the socket or remove the node for the next 100ms.

Also, when node3 has fDisconnect=true and is lingering, it is already able to initiate a new connection to node0, which still believes that node3 is connected. As node3 is a MN, it will be disconnected immediately due to duplicate MNAUTH and a few ms later the old connection is removed as well, leaving it with 0 connections to node3.

The latest commit should fix this.

**Comment by UdjinM6** [comment]
Your fix makes sense and gitlab is green but tests still fail sometimes for me the same way they did before when I run them like that:
```
./test/functional/test_runner.py p2p-acceptblock.py llmq-simplepose.py --dashd-arg="-socketevents=select"
```
I guess the reason is that we only run `DisconnectNodes()` once per second now (and not on rpcs like before). Reverting 30e4afae002d0ba24cc930075252af33e3a7c623 helps.

**Comment by codablock** [comment]
@UdjinM6 Ok, I finally decided to revert https://github.com/dashpay/dash/commit/30e4afae002d0ba24cc930075252af33e3a7c623. I'll later re-investigate the performance issues I had with the disconnect loop and then decide if a new/proper fix is needed.

**Pull request merged**